### ISSUE TITLE
support non-blocking files

### ISSUE THREAD

**Issue by codefromthecrypt** [tetratelabs/wazero#1500]
**Is your feature request related to a problem? Please describe.**

Currently, we support non-blocking stdio and sockets, but not files. This leads to an awkward and unnecessary [skip in go's tests](https://github.com/golang/go/blob/5f695f7072848e425289ca94bc460da813de66da/src/runtime/internal/wasitest/nonblock_test.go#L41).

**Describe the solution you'd like**

Go supports non-blocking access and we support native access to files, so we should be able to implement this. If there are any caveats about windows, support can begin with a limited set of platforms.

**Describe alternatives you've considered**

continue to not do this

**Additional context**

our file type already has a SetNonblock setting, so that's a start. We also have a separate file implementation for raw os-backed files. So, we should be able to figure out why we aren't implementing non-blocking access for os files, then figure out how to implement it.

**Comment by evacchi** [comment]
a quick summary of what's required to make this works.

In theory, we can easily enable SetNonblock() for regular files, but effectively it would only work on *NIX. In fact, the `syscall.SetNonblock()` fun on Windows is a no-op.

Now, even if we did support that only on *NIX, we should still fix poll_oneoff; because as much as it would still work without it (read() returns EAGAIN and you could just retry after a little), existing code may still assume that `select` works.

so:

- we can enable syscall.Setnonblock() on regular files
- it won't work on Windows unless we explicitly implement something
- read() now does not block and may return EAGAIN
- we should fix poll_oneoff because it is currently special-cased for stdin.

now the good news about `poll_oneoff` is that there actually _is_ a `select()` implementation that works on Windows
... the bad news it only works on sockets. Better than nothing.

So in short:

- regular files:
- easy to set the flag for nonblocking
- requires reworking poll_oneoff (but this can be done in a separate PR)
  - won't work for free on Windows
  - would work almost for free on Windows+Sockets (via WinSock select)

**Comment by codefromthecrypt** [comment]
some trouble in HTTP land. I began troubleshooting with @chriso but I think we should all be able to enjoy a weekend and I don't think this will quickly resolve. https://github.com/tetratelabs/wazero/pull/1503

**Comment by codefromthecrypt** [comment]
so next step is to really look carefully at windows and sockets, and find some real-world wasm (very hard to find, like [this](https://gist.github.com/chriso/6c71e968ef1002981a6ff46ceaa39ee3#file-http-go-L59-L78) and [this](https://github.com/enarx/codex/blob/d5c9144cfa663e612a2bc7d9104cc65bea14955c/demos/chat-client/rust/src/main.rs#L30-L33)) which can verify assumptions.

**Comment by evacchi** [comment]
as for PR #1502, this only enables non-blocking I/O for files on *NIX. 

Interesting notes, really on *regular* files poll always returns OK, except that then read() may still return EAGAIN l https://www.remlab.net/op/nonblock.shtml. (In the tests we are using pipes). On Windows Pipes _do exist_ but they require quite a bit of boilerplate to initialize, to the point there are literally go libraries that only do that https://github.com/gesellix/go-npipe https://github.com/natefinch/npipe

Summary for further work on Windows.

- Windows technically only does async I/O, but as far as I have understood it works with a callback-based mechanism, that doesn't play exactly the same as select/poll
- Go's built-in impl for `syscall.SetNonblock()` is "LOLJK":
   ```
    func SetNonblock(fd Handle, nonblocking bool) (err error) {
        return nil
    }
```
- Select is possible for sockets on Windows; actually _only_ on sockets, because WinSock provides that https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-select ; Notice that [even Python provides `select` only for sockets _on Windows_](https://docs.python.org/3/library/select.html)

Long story short, it should be feasible to provide select, and thus poll_oneoff for sockets on Windows too, but it will require some rework of the current `poll_oneoff` impl, which is really ad-hoc for stdin :^)

**Comment by codefromthecrypt** [comment]
Thanks for the update. Really good info!

Aside: I think one thing we need before closing this also, is a gotip version of the zig non-blocking test, as it isn't intuitive how to set it up. the question came up on slack by @tegk.

**Comment by evacchi** [comment]
some further work in #1517 on unix. This should address most issues on this platform (with some love probably still needed for poll_oneoff), while work on Windows needs the most attention

**Comment by chriso** [comment]
Related: https://github.com/tetratelabs/wazero/issues/1538

**Comment by evacchi** [comment]
#1570 improved support to pipes on Windows. I think that if we fix sockets (just use WinSock select) and we tie everything together in poll_oneoff, we are getting closer to resolving this!

**Comment by shynome** [comment]
hi, is there any api to set nonblock 

this make me fall down again.  :(

**Comment by paralin** [comment]
Here is poll_oneoff: https://github.com/tetratelabs/wazero/blob/v1.9.0/imports/wasi_snapshot_preview1/poll.go#L52

The current logic seems to return success immediately for any FdRead other than stdin:

```
case wasip1.EventTypeFdRead:
  fd := int32(le.Uint32(argBuf))
  if fd < 0 {
    return sys.EBADF
  }
  if file, ok := fsc.LookupFile(fd); !ok {
    evt.errno = wasip1.ErrnoBadf
    writeEvent(outBuf[outOffset:], evt)
    nevents++
  } else if fd != internalsys.FdStdin && file.File.IsNonblock() {
    writeEvent(outBuf[outOffset:], evt)
    nevents++
  } else {
    // if the fd is Stdin, and it is in blocking mode,
    // do not ack yet, append to a slice for delayed evaluation.
    blockingStdinSubs = append(blockingStdinSubs, evt)
  }
```

and then below hardcodes a call to Poll on the Stdin file descriptor:

```
	// Wait for the timeout to expire, or for some data to become available on Stdin.
	if stdinReady, errno := stdin.File.Poll(fsapi.POLLIN, int32(timeout.Milliseconds())); errno != 0 {
		return errno
	} else if stdinReady {
		// stdin has data ready to for reading, write back all the events
		for i := range blockingStdinSubs {
			evt := blockingStdinSubs[i]
			evt.errno = 0
			writeEvent(outBuf[nevents*32:], evt)
			nevents++
		}
	}
```

I see other logic for implementing Poll for network sockets: https://github.com/tetratelabs/wazero/blob/v1.9.0/internal/sysfs/sock_unix.go#L21

But this seems to be hardcoding 0 for the timeout on poll() - it doesn't seem like there is anywhere else calling Poll other than the hardcoded logic above.

I have to then conclude that:

- The sock poll() is currently not used
- Only stdin will properly do async i/o, other file descriptors will return ready immediately regardless of Poll() implementation
- We don't do async i/o on custom FS Read() / Poll() implementations either

The best way to solve this would be to modify the loop. The current logic is (roughly):

- for each subscription we want to wait for:
   - if it is a clock event: update the clock, update the current minimum timeout value
   - if it is a fd read event: 
     - return err if fd not exist
     - if the fd is stdin or IsNonblock returns false: queue the fd to blockingStdinSubs
          - **this seems like a bug**: we will call Poll on stdin instead of on the correct fd for IsNonblock()==false
              - won't this trigger a EFAULT every time we try to read from a IsNonblock() fd like a socket?
          - current: `else if fd != internalsys.FdStdin && file.File.IsNonblock()` => write success event
          - shouldn't this be `else if fd != internalsys.FdStdin || !file.File.IsNonblock()` ?
          - and below that, **we need to Poll() on multiple pollables at a time**
   - if it is a fd write event, return success immediately (dont defer writing)
- if we wrote a response (error or success) for all subscriptions, sleep for timeout, then return.
- otherwise wait for stdin to Poll() for the timeout. **this is wrong, we should Poll() the affected fds**

I think as far as I can see the best way to implement this is:

- If the timeout is very low (less than 10ms?) then:
   - Call Poll() on the first fd with the timeout set.
   - Call Poll() on the remaining fds with timeout=0.
   - Write events for the results back to wasi
- If the timeout is longer, then:
   - Call Poll() on the fds immediately with timeout=0
   - If any are ready, write events back, return right away.
   - Otherwise: create a channel + goroutine for each of the blocking fds
   - Within the goroutine call Poll() with the timeout set, then close the channel when done.
       - To avoid concurrent Poll calls we could set a lower inner timeout and loop 
   - Within the outer pollOneoffFn select{} on either context canceled, time.After(timeout), or channel closures
   - With some clever design, we could re-use channels for fds (signaling unblock) to speed this up.

This seems like the correct way to get non-blocking I/O working as expected, thoughts?

**Comment by paralin** [comment]
I think there is an opportunity to use a much more Go-centric way to wait for multiple fds to be ready. While wasi expects poll_oneoff to return events when either a) a fd is ready to yield data or b) the timeout has elapsed, we can internally implement this behavior in a variety of ways. I would suggest leveraging the Go scheduler to wait on multiple fds and/or virtual File interface objects to be ready through select{} statements and channels, for example, as opposed to emulating Poll() with blocking semantics.

**Comment by shynome** [comment]
it looks like `internal/sysfs.OpenFSFile` still don't support non-blocking files now
```go
	fsc := wazero.NewFSConfig()
	fsc = fsc.WithFSMount(fsnet.New("/dev/tcp/"), "/dev/tcp")
```
----
update: I add word `still`. sorry for my bad english
my issue at https://github.com/wazero/wazero/issues/2208
I have impl nonblocker interface but still block, orz. I give up it now, someday resume it

**Comment by ncruces** [comment]
@paralin do you have a bit to look into this?

**Comment by paralin** [comment]
@ncruces @shynome wazero only checks Poll on stdin and this limitation was present before the recent PR. I realize now that @shynome is asking that we support Poll on any file descriptor, not just on stdin. To do that, it's a new "feature" in wazero (so this was not a regression in the recent PR adding PollableFile). I opened #2481 to add this support, with the cavet in #2480 that we don't have a clean way to call Poll on multiple file descriptors simultaneously at the moment (which is why the limitation to just stdin polling probably existed before).

While investigating this I found a separate issue about SetNonblock and filed #2479 to solve that.

**Comment by paralin** [comment]
@shynome However, your `fs.File` from fsnet also needs to implement the polling interface for wazero to know how to poll it. wazero can't assume anything about an arbitrary `fs.File` - the file itself needs to declare poll support.

Your `fs.File` wrapping `net.Conn` should implement `experimental/sys.Pollable`:

```go
import experimentalsys "github.com/tetratelabs/wazero/experimental/sys"

func (f *yourFile) Poll(flag experimentalsys.Pflag, timeoutMillis int32) (ready bool, errno experimentalsys.Errno) {
    if flag == experimentalsys.POLLIN {
        // Use net.Conn.SetReadDeadline to implement poll semantics
        // ...
    }
    return false, experimentalsys.ENOTSUP
}
```

And optionally for non-blocking mode:

```go
func (f *yourFile) IsNonblock() bool { ... }
func (f *yourFile) SetNonblock(enable bool) experimentalsys.Errno { ... }
```

**Comment by paralin** [comment]
To clarify, this was NOT a regression from PR 2476, before that PR:

- `fsFile.IsNonblock()` returned `false` (hardcoded)
- `fsFile.Poll()` returned `ENOSYS` (hardcoded)
- `fsFile.SetNonblock()` returned `ENOSYS` (hardcoded)

There was no forwarding to the underlying `fs.File`. So fsnet's files never worked with poll_oneoff.

The fsnet `File` struct needs to implement `Pollable` at minimum.

**Comment by paralin** [comment]
@ncruces I split it into two PRs: #2481 and #2479

**Comment by shynome** [comment]
@paralin thanks for your suggestion. my issue at https://github.com/wazero/wazero/issues/2208, maybe test case help
I have impl nonblocker interface but still block, orz. I give up it now, someday resume it

**Comment by paralin** [comment]
@shynome it will work when:

1) #2479 is merged
2) #2481 is merged
3) you implement Poll on your file according to what I wrote earlier in this thread.

have a look! you could try checking out a clone of wazero and merging those two in and try it.

**Comment by shynome** [comment]
@paralin Thank you. Actually, I had already tried it when I left the previous comment, but it still failed. So I gave up for now, QAQ.

**Comment by paralin** [comment]
Relevant work in tinygo: https://github.com/tinygo-org/tinygo/commit/d01a932201bb4059eb6c84f1ee4e7e5e84f70040
